[operator] Add `Care` reconciler to `Garden` controller. (#8158)
gardener/gardener: docs/concepts/operator.md

- `VirtualGardenAPIServerAvailable`: The `/healthz` endpoint of the garden's `virtual-garden-kube-apiserver` is called and considered healthy when it responds with `200 OK`.

If all checks for a certain condition are succeeded, then its `status` will be set to `True`.
Otherwise, it will be set to `False` or `Progressing`.

If at least one check fails and there is threshold configuration for the conditions (in `.controllers.gardenCare.conditionThresholds`), then the status will be set:

- to `Progressing` if it was `True` before.
- to `Progressing` if it was `Progressing` before and the `lastUpdateTime` of the condition does not exceed the configured threshold duration yet.
- to `False` if it was `Progressing` before and the `lastUpdateTime` of the condition exceeds the configured threshold duration.

The condition thresholds can be used to prevent reporting issues too early just because there is a rollout or a short disruption.
Only if the unhealthiness persists for at least the configured threshold duration, then the issues will be reported (by setting the status to `False`).

> 

Reviewer: Please check https://github.com/gardener/gardener/blob/master/docs/concepts/gardenlet.md#shoot-controller and follow this structure:

# Garden Controller
## "Main" Reconciler
## "Care" Reconciler

Author: I changed the levels of a couple of other headings too, that they fit into the schema.

The controller maintains the `Reconciled` condition which indicates the status of an operation.

#### [`GardenCare` Controller](../../pkg/operator/controller/garden/care)

This reconciler performs four "care" actions related to `Garden`s.

It maintains four conditions and performs the following checks:

> - `SystemComponentsHealthy`: The conditions of the `ManagedResource`s applied to the runtime cluster are checked (e.g., `ResourcesApplied`).
> - `VirtualGardenComponentsHealthy`: The conditions of the `ManagedResource`s applied to the virtual cluster are checked (e.g., `ResourcesApplied`).

Reviewer 2: ```suggestion
- `RuntimeComponentsHealthy`: The conditions of the `ManagedResource`s applied to the runtime cluster are checked (e.g., `ResourcesApplied`).
- `VirtualComponentsHealthy`: The conditions of the `ManagedResource`s applied to the virtual cluster are checked (e.g., `ResourcesApplied`).
```

To reflect the terms used in the `garden` specification.

- `VirtualGardenAPIServerAvailable`: The `/healthz` endpoint of the garden's `virtual-garden-kube-apiserver` is called and considered healthy when it responds with `200 OK`.

> - `VirtualGardenControlPlaneHealthy`: The control plane is considered healthy when the respective `Deployment`s (for example `virtual-garden-kube-apiserver`,`virtual-garden-kube-controller-manager`), and `Etcd`s (for example `virtual-garden-etcd-main`) exist and are healthy.

Reviewer 2: ```suggestion
```

We thinks it's rather better to drop this condition and move the checks to `VirtualComponentsHealthy`. WDYT?
You probably wanted to align to what we already do for shoots but there we have actual workload in the cluster that we need to check, in contrast to the virtual garden cluster where we'll never have any workload. Hence, checking the `ManagedResource`s and putting the result into a dedicated condition doesn't make much sense.

> If all checks for a certain conditions are succeeded, then its `status` will be set to `True`.

Reviewer 2: ```suggestion
If all checks for a certain condition are succeeded, then its `status` will be set to `True`.
```

> Otherwise, it will be set to `False`.

Reviewer 2: ```suggestion
Otherwise, it will be set to `False` or `Progressing`.
```
